File: Leetcode/2360. Longest Cycle in a Graph.cpp
class Solution {
public:
    int res=-1;
    void dfs(int len,int node,vector<int>& visited,vector<int>& cp,vector<int>& edges){
        visited[node]=1;
        cp[node]=len+1;
        if(edges[node]!=-1){
            if(!visited[edges[node]]){
                //continue dfs
                dfs(len+1,edges[node],visited,cp,edges);
            }else if(cp[edges[node]]!=0){
                //visited and cycle detected
                res=max(res,cp[node]-cp[edges[node]]+1);
            }
        }
        cp[node]=0; //backtracking again
        return;
    }
    int longestCycle(vector<int>& edges) {
        int n=edges.size();
        vector<int> visited(n,0);
        vector<int> cp(n,0);
        for(int i=0;i<n;i++){
            if(!visited[i]){
                dfs(0,i,visited,cp,edges);
            }
        }
        return res;
    }
};

File: Leetcode/2360. Longest Cycle in a Graph.py
class Solution:
    def __init__(self):
        self.res=-1
    def longestCycle(self, edges: List[int]) -> int:
        n=len(edges)
        visited=[0]*n
        cp=[0]*n
        for i in range(n):
            if visited[i]==0:
                self.dfs(0,i,visited,cp,edges)
        return self.res
    def dfs(self,l,node,visited,cp,edges):
        visited[node]=1
        cp[node]=l+1
        nbr = edges[node]
        if nbr!=-1:
            if visited[nbr]!=1:
                self.dfs(l+1,nbr,visited,cp,edges)
            elif cp[nbr]!=0:
                #cycle detected
                self.res=max(self.res,cp[node]-cp[nbr]+1)
        cp[node]=0 #backtrack and going back
        return
